src/celeritas/optical/OpticalGenData.hh

```diff
@@ -0,0 +1,146 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/OpticalStepData.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/data/Collection.hh"
+#include "corecel/data/StackAllocatorData.hh"
+#include "celeritas/Quantities.hh"
+#include "celeritas/Types.hh"
+#include "celeritas/optical/CerenkovData.hh"
+#include "celeritas/optical/OpticalDistributionData.hh"
+#include "celeritas/optical/OpticalPropertyData.hh"
+#include "celeritas/optical/ScintillationData.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Immutable problem data for generating optical photon distributions.
+ */
+template<Ownership W, MemSpace M>
+struct OpticalGenParamsData
+{
+    //// DATA ////
+
+    bool cerenkov{false};  //!< Whether Cerenkov is enabled
+    bool scintillation{false};  //!< Whether scintillation is enabled
+    real_type stack_capacity{0};  //!< Distribution data stack capacity
+
+    //// METHODS ////
+
+    //! True if all params are assigned
+    explicit CELER_FUNCTION operator bool() const
+    {
+        return (cerenkov || scintillation) && stack_capacity > 0;
+    }
+
+    //! Assign from another set of data
+    template<Ownership W2, MemSpace M2>
+    OpticalGenParamsData& operator=(OpticalGenParamsData<W2, M2> const& other)
+    {
+        CELER_EXPECT(other);
+        cerenkov = other.cerenkov;
+        scintillation = other.scintillation;
+        stack_capacity = other.stack_capacity;
+        return *this;
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Pre-step data needed for generating optical photon distributions.
+ */
+struct OpticalPreStepData
+{
+    units::LightSpeed speed;  //!< Pre-step speed
+    Real3 pos{};  //!< Pre-step position
+    real_type time{};  //!< Pre-step time
+
+    //! Check whether the data are assigned
+    explicit CELER_FUNCTION operator bool() const
+    {
+        return speed > zero_quantity();
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Optical photon distribution data and primaries.
+ */
+template<Ownership W, MemSpace M>
+struct OpticalGenStateData
+{
+    //// TYPES ////
+
+    template<class T>
+    using StateItems = StateCollection<T, W, M>;
+    using DistributionStackData
+        = StackAllocatorData<OpticalDistributionData, W, M>;
+
+    //// DATA ////
+
+    // Pre-step data for generating optical photon distributions
+    StateItems<OpticalPreStepData> step;
+
+    // Buffers of distribution data for generating optical primaries
+    DistributionStackData cerenkov;
+    DistributionStackData scintillation;
```

Comment:
I need to think a little more about how we're going to ensure reproducibility if we're using a stack allocator to store the distribution data over possibly multiple steps... maybe these should just be state-sized collections since there will be max one distribution per track per generating process at each step, and we can move the data to a separate larger buffer for long-term storage (like we do for the track initializers)? Or maybe even just store them directly in the larger buffer and then compact them with a `remove_if` at each step?

Reply:
Ok, I'm trying out the latter...

Reply:
Good point about reproducibility. The only other thing I can think would be to make a sparse version of that by also storing the originating thread/track ID... but `remove_if` sounds more efficient.